### PR TITLE
Add message type and space id to history

### PR DESCRIPTION
**Pull request by marcin-cebo**

### CODE REVIEW

File: src/integrationTest/java/com/pubnub/api/integration/HistoryIntegrationTest.java [hunk 1]
```diff
@@ -164,6 +170,39 @@ public void testFetchSingleChannel_Meta() throws PubNubException {
...
+        PNPublishResult pnPublishResult = pubNub.publish()
+                .message(message)
+                .channel(expectedChannelName)
+                .messageType(new MessageType(expectedMessageTypeValue))
+                .spaceId(new SpaceId(expectedSpaceIdValue))
+                .sync();
+
+        assertNotNull(pnPublishResult.getTimetoken());
+
+        pause(3);
```
[the remark this answers is not in the document]
Comment (marcin-cebo) [reply src/integrationTest/java/com/pubnub/api/integration/HistoryIntegrationTest.java:189]: Make sense. I will change it.

File: src/integrationTest/java/com/pubnub/api/integration/HistoryIntegrationTest.java [hunk 2]
```diff
@@ -164,6 +170,39 @@ public void testFetchSingleChannel_Meta() throws PubNubException {
 
     }
 
+    @Test
+    public void testFetchSingleChannel_SpaceId_MessageType() throws PubNubException {
+        final String expectedChannelName = random();
+        final String message = "message";
+        final String expectedSpaceIdValue = "mySpace";
+        final String expectedMessageTypeValue = "myMessageType";
```
[the remark this answers is not in the document]
Comment (marcin-cebo) [reply src/integrationTest/java/com/pubnub/api/integration/HistoryIntegrationTest.java:178]: ok

File: src/main/java/com/pubnub/api/endpoints/FetchMessages.java [hunk 3]
```diff
@@ -226,12 +210,52 @@ private JsonElement processMessage(JsonElement message) throws PubNubException {
...
+    private Map<String, String> extendRequestParams(Map<String, String> params) {
+        params.put(MAX_QUERY_PARAM, String.valueOf(maximumPerChannel));
+        params.put(INCLUDE_UUID_QUERY_PARAM, Boolean.toString(includeUUID));
+        if (start != null) {
+            params.put(START_QUERY_PARAM, Long.toString(start).toLowerCase());
+        }
+        if (end != null) {
+            params.put(END_QUERY_PARAM, Long.toString(end).toLowerCase());
+        }
+        String yes = Boolean.toString(true);
```
[the remark this answers is not in the document]
Comment (marcin-cebo) [reply src/main/java/com/pubnub/api/endpoints/FetchMessages.java:251]: We need it in 
params.put(INCLUDE_META_QUERY_PARAM, yes);
but I can inline it.

File: src/test/java/com/pubnub/api/models/consumer/history/PNFetchMessageItemTest.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+class PNFetchMessageItemTest {
+    private PNFetchMessageItem objectUnderTest;
+
+
+    @Test
+    void when_includeMessageType_is_false_then_getMessageType_should_throw_exception() {
+        boolean includeMessageType = false;
+        Object spaceId = null;
+        Object pnMessageType = null;
+        Object userMessageType = null;
```
[the remark this answers is not in the document]
Comment (marcin-cebo) [reply src/test/java/com/pubnub/api/models/consumer/history/PNFetchMessageItemTest.java:22]: It's by mistake. I will fix that.

File: src/integrationTest/java/com/pubnub/api/integration/PublishIntegrationTests.java [hunk 5]
```diff
@@ -84,6 +86,24 @@ public void testPublishMessageWithMessageTypeAndSpaceId() throws PubNubException
...
+                .sync();
+
+
+        for (PNFetchMessageItem messageItem : fetchMessagesResult.getChannels().get(expectedChannel)) {
+            assertNotNull(messageItem.getMessage());
+            assertNotNull(messageItem.getTimetoken());
+            assertNull(messageItem.getMeta());
+            assertNull(messageItem.getActions());
+            assertNotNull(messageItem.getMessageType());
+            assertNotNull(messageItem.getSpaceId());
```
[the remark this answers is not in the document]
Comment (marcin-cebo) [reply src/integrationTest/java/com/pubnub/api/integration/PublishIntegrationTests.java:105]: I can delete:
            assertNotNull(messageItem.getMessage());
            assertNotNull(messageItem.getTimetoken());
            assertNull(messageItem.getMeta());
            assertNull(messageItem.getActions());
leaving 
            assertEquals(expectedSpaceId, messageItem.getSpaceId());
            assertEquals(expectedMessageType, messageItem.getMessageType());
to fulfil what stated in test name.
What do you think?

File: src/integrationTest/java/com/pubnub/api/integration/HistoryIntegrationTest.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (marcin-cebo) [reply src/integrationTest/java/com/pubnub/api/integration/HistoryIntegrationTest.java:189]: `pause(3)` is commonly used in all test in HistoryIntegrationTest.java
in java code with `await` looks a bit strange

```
        Awaitility.await()
                .pollInterval(Duration.ofSeconds(1))
                .pollDelay(Duration.ofSeconds(1))
                .atMost(Duration.ofSeconds(5))
                .untilAsserted(() -> {
                    fetchMessagesResult[0] = pubNub.fetchMessages()
                            .channels(Collections.singletonList(expectedChannelName))
                            .maximumPerChannel(25)
                            .sync();

                    assertNotNull(fetchMessagesResult[0]);
                    assertThat(fetchMessagesResult[0].getChannels().size(), greaterThan(1));
                });
```

we can discuss this?

File: src/test/java/com/pubnub/api/endpoints/FetchMessagesTest.java [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+    void when_includeMessageType_equal_true_is_specified_explicitly_in_api_call_then_request_should_contain_includeMessageType_request_param_set_to_true() throws PubNubException {
+        Map<String, String> baseParams = getBaseParams();
+        ArgumentCaptor<Map<String, String>> extendedParamsCaptor = ArgumentCaptor.forClass(Map.class);
+        objectUnderTest.channels(Arrays.asList("channel"));
+        objectUnderTest.includeMessageActions(false);
+        objectUnderTest.includeMeta(true);
+        objectUnderTest.includeMessageType(true);
+        objectUnderTest.includeSpaceId(false);
+
+        objectUnderTest.doWork(baseParams);
```
[the remark this answers is not in the document]
Comment (marcin-cebo) [reply src/test/java/com/pubnub/api/endpoints/FetchMessagesTest.java:63]: After such modification will it be still unitTest? I wanted it to be unitTest. 
Let's talk.

File: src/main/java/com/pubnub/api/models/consumer/history/PNFetchMessageItem.java [hunk 7]
```diff
@@ -19,14 +25,44 @@ public class PNFetchMessageItem {
...
+     *
+     * @return MessageType of PNFetchMessageItem
+     * @throws PubNubRuntimeException if includeMessageType is set to false
+     */
+    public MessageType getMessageType() {
+        if (includeMessageType) {
+            return new MessageType(PNMessageType.valueByPnMessageType(pnMessageType), userMessageType);
         } else {
-            return Integer.parseInt(messageType);
+            throw PubNubRuntimeException.builder().pubnubError(PNERROBJ_TO_GET_MESSAGE_TYPE_INCLUDEMESSAGETYPE_SHOULD_NOT_BE_SET_TO_FALSE).build();
```
[the remark this answers is not in the document]
Comment (marcin-cebo) [reply src/main/java/com/pubnub/api/models/consumer/history/PNFetchMessageItem.java:65]: Good question. I forgot to discuss this. I thought that getMessageType is kind of special and throwing exception will be   more informative. We can throw exception also from getSpaceId but then what about `includeMeta` should it also throw exception?

Let's talk

[the remark this answers is not in the document]
Comment (kleewho) [reply src/main/java/com/pubnub/api/models/consumer/history/PNFetchMessageItem.java:65]: I would rather not throw. Null is still a valid information which is not that very exceptional in my opinion

File: src/integrationTest/java/com/pubnub/api/integration/HistoryIntegrationTest.java [hunk 8]
```diff
@@ -146,22 +158,57 @@ public void testFetchSingleChannel_Meta() throws PubNubException {
...
             assertNotNull(messageItem.getMeta());
             assertNull(messageItem.getActions());
         }
+    }
+
+    @Test
+    public void testFetchSingleChannel_SpaceId_MessageType() throws PubNubException {
+        final String expectedChannelName = random();
+        final String message = "message";
+        final String expectedSpaceIdValue = "mySpace";
```
Comment (kleewho) [inline src/integrationTest/java/com/pubnub/api/integration/HistoryIntegrationTest.java:184]: You could do the same trick with SpaceId as with the MessageType and use the type across the test

**Review by kleewho — APPROVED** [review]

**Pull request merged**